design/design.md

```diff
@@ -0,0 +1,251 @@
+Hash functions evolve, old ones get broken (SHA-1), and so our ALBA
+implementation shouldn't get stuck with an inefficient / broken hash function.
+Moreover, the user might want to instantiate ALBA with a PRF (e.g. AES) instead
+of a hash function modeled as a random oracle for better theoretical guarantees.
+With Rust's advanced trait system, making the hash function generic should be
+easy, and the runtime code performance should be fast due to compile time
+function resolution.
+
+We will provide an example implementing a hash function with the right interface
+for ALBA using, for example, BLAKE underneath, but with a warning that
+it is provided for demonstration / prototyping only and should not be used in
+production.
+
+One reason is that when ALBA is used alongside other sub-protocols that also
+need hashing, it is the user's responsibility to implement
+[domain separation](https://en.wikipedia.org/wiki/Domain_separation)
+to provide the different
+sub-protocols with independent hash functions.
+Otherwise, the underlying assumptions needed for security might break and the
+larger protocol can become insecure
+(see [here](https://link.springer.com/chapter/10.1007/978-3-030-45724-2_1)).
+
+Additionally, different hash functions (e.g., SHA and BLAKE) shouldn't mix and
+match.
+The following example is contrived, but it demonstrates why using different
+hash function constructions within the same protocol can be insecure.
+Suppose $H$ and $H'$ are two distinct random looking hash functions.
+You might expect that the function $G$ given by $G(x) = H(x) \oplus H'(x)$,
+where $\oplus$ is the bitwise xor, is also random looking, but it's not true.
+Let $H$ be a hash function modeled as a random oracle and define
+$H(x) = \sim H'(x)$,
+where $\sim$ is bitwise not; then $H'$ is also random looking
+but $G(x) = 111...111$ for all $x$.
+Even if we use domain separation for $H$ and $H'$, we cannot guarantee that
+$G(x) = H(0, x) \oplus H'(1, x)$ is random looking.
+Define $H'(b, x) = H(\sim b, x)$; then $G(x) = 000...000$ for all $x$.
+
+## Generic data elements
+Similar to the above, we should also let data elements have a generic type.
+For example, the ALBA implementation should work with a signature type that
+exposes the underlying data as a `u8` slice.
+The exact interfaces will need to be determined.
+
+## Mathematical rigor and production-readiness
+At this point, it is not our goal to make the implementation 100% correct and
+secure.
+We need to roll out a prototype so that our primary clients (Peras, etc) can
+incorporate it in their prototypes, and also so that the external community can
+start playing with it as well.
+This means that there might be tiny rounding errors in parameter derivation due
+to the use of floats, and there might be tiny errors when using a uniformly
+random hash output to sample some distribution needed for ALBA
+(e.g., Bernoulli, Uniform, Binomial, Poisson).
```

Comment:
This is already taken into account in the paper's appendix, so define correctness.